# stDCL

[![PyPI - Python Version](https://img.shields.io/pypi/v/stDCL)](https://pypi.org/project/stDCL/)
[![License](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.10968451.svg)](https://zenodo.org/records/10968451)

`stDCL` is a Python package containing tools for identifing spatial domains from spatial transcriptomics data based on a dual graph contrastive learning method.

- [Overview](#overview)
- [System Requirements](#system-requirements)
- [Installation Guide](#installation-guide)
- [Usage](#Usage)
- [Data Availability](#data-availability)
- [License](#license)


# Overview
Recent advances in spatial transcriptomics have enabled simultaneous preservation of high-throughput gene expression profiles and spatial context, enabling the high-resolution exploration of distinct regional characterization in brain tissue. However, the complex spatial structure of brain poses challenges in capturing spatial heterogeneity effectively and deciphering internal regulatory mechanisms accurately. Here, we develop stDCL, a dual graph contrastive learning method to identify spatial domains from spatial transcriptomics data. stDCL adaptively incorporates gene expression data and spatial information via a graph embedding autoencoder, thereby preserving critical information within latent embedding representations. Then, we propose dual graph contrastive learning to train the model, ensuring that the latent embedding representation closely resembles the actual spatial distribution and exhibits cluster similarity. Benchmarking stDCL against other state-of-the-art clustering methods using cerebral cortex datasets demonstrates its superior accuracy and effectiveness in identifying spatial domains. Our analysis of the imputation matrices generated by stDCL reveals its capability to reconstruct spatial hierarchical structures and refine differential expression assessments. Furthermore, we illustrate the central role of stDCL in elucidating gene regulation, spatial heterogeneity, and developmental patterns in the brain. Additionally, stDCL can also be instrumental in annotating disease-associated cellular phenotypes in Alzheimer's disease, unraveling the regulatory mechanisms.


![1811713010487_ pic](https://github.com/Philyzh8/stDCL/assets/65069252/f15928c1-59c2-4936-82f2-be39a3426456)

# System Requirements
## Hardware requirements
`stDCL` package requires only a standard computer with enough RAM to support the in-memory operations.

## Software requirements
### OS Requirements
This package is supported for *Linux*. The package has been tested on the following systems:
+ Linux: Ubuntu 18.04

### Python Dependencies
`stDCL` mainly depends on the Python scientific stack.
```
numpy
scipy
torch
scikit-learn
pandas
scanpy
```
For specific setting, please see <a href="https://github.com/Philyzh8/stDCL/blob/master/requirements.txt">requirements</a> or <a href="https://github.com/Philyzh8/stDCL/blob/master/environment.yaml">environment</a>.

# Installation Guide:

### Install from PyPi

```
$ conda create -n stDCL_env python=3.8.15
$ conda activate stDCL_env
$ pip install -r requirements.txt
$ pip install stDCL
```

### Install from Conda

```
$ conda env create -f environment.yaml
```

# Usage
:page_facing_up: `stDCL` is a dual graph contrastive learning method for identifing spatial domains, which can be used to:
+ DLPFC. The example can be seen in the <a href="https://github.com/Philyzh8/stDCL/blob/master/tutorial/run_dlpfc.py">run_dlpfc.py</a>.
+ MERFISH. The example can be seen in the <a href="https://github.com/Philyzh8/stDCL/blob/master/tutorial/run_merfish.py">run_merfish.py</a> and <a href="https://github.com/Philyzh8/stDCL/blob/master/tutorial/run_merfish_brain.py">run_merfish_brain.py</a>.
+ osmFISH. The example can be seen in the <a href="https://github.com/Philyzh8/stDCL/blob/master/tutorial/run_osm.py">run_osm.py</a>.
+ STARmap. The example can be seen in the <a href="https://github.com/Philyzh8/stDCL/blob/master/tutorial/run_starmap.py">run_starmap.py</a> and <a href="https://github.com/Philyzh8/stDCL/blob/master/tutorial/run_starmap_plus.py">run_starmap_plus.py</a>.

:page_facing_up: We also provide <a href="https://github.com/Philyzh8/stDCL/tree/master/step-by-step">step-by-step</a> tutorials to analyze spatial transcriptomic datasets.

# Arguments

|    Parameter    | Introduction                                                 |
| :-------------: | ------------------------------------------------------------ |
|    dataset     | Contains spatial transcriptomics gene expression matrix, coordinate information, and other information. By default, it is assumed that genes are represented by columns and samples are represented by rows.|
|  datatype   | Different spatial transcriptomics data types, such as ‘10x’, ‘osmFISH’, ‘MERFISH’, etc.                                    |
|  n_clusters   | The true number of clusters in the data or the pre-set number of clusters                                    |
| prelr | Learning rate for model pre-training                                   |
| lr | Learning rate for model training                                   |
| pre_epochs | Number of pre-training epochs                                    |
|     epochs     | Number of training epochs                                    |
|     noise     | Gaussian noise added during data augmentation                                  |
|     lambda1-3     | The weight coefficient assigned to each loss                                 |
|     input     | Number of  highly variable genes selected                                 |
|     latent_dim    | Output dimension of the hidden layer                                 |
|     n_neighbors    | For spatial location information, the number of neighboring spots for each spot                                 |
|     n_neighbors_gene    | For gene expression information, the number of neighboring spots for each spot                                 |
|     clustertype    | The clustering method used, including ‘mclust’, ‘leiden’, ‘louvain’                              |
|     radius    | When the clustering results are redefined, each spot is reassigned to the radius of its respective neighborhood                              |



# Data Availability

The real data sets we used can be download in <a href="https://zenodo.org/records/10968451">data</a>.

# License

This project is covered under the <a href="https://github.com/Philyzh8/stDCL/blob/master/LICENSE">**MIT License**</a>.


